.teamcity/_self/projects/WebApp.kt

```diff
@@ -222,6 +229,56 @@ object BuildDockerImage : BuildType({
 				EOF
 			"""
 		}
+
+		// Conditions don't seem to support and/or, so we do this in a separate step.
+		// Essentially, UPDATE_BASE_IMAGE_CACHE will remain false by default, but
+		// if we're on trunk and get WEBPACK_CACHE_INVALIDATED set by the docker build,
+		// then we can flip it to true to trigger the cache rebuild.
+		script {
+			name = "Set cache update"
+			conditions {
+				equals("env.WEBPACK_CACHE_INVALIDATED", "true")
+				equals("teamcity.build.branch.is_default", "true")
+			}
+			scriptContent = """
+				echo "##teamcity[setParameter name='UPDATE_BASE_IMAGE_CACHE' value='true']"
+			"""
+		}
+
+		// This updates the base docker image when the webpack cache invalidates.
+		// It does so by re-using the layers already generated above, and simply
+		// copying the .cache directory as a new layer into the base image. On
+		// trunk, this will update the latest base image for future builds.
+		//
+		// Runs after everything else to avoid blocking the deploy system or calypso.live.
+		dockerCommand {
+			name = "Rebuild cache image"
+			conditions {
+				equals("UPDATE_BASE_IMAGE_CACHE", "true")
+			}
+			commandType = build {
+				source = file {
+					path = "Dockerfile"
+				}
+				namesAndTags = "registry.a8c.com/calypso/base:%base_image_publish_tag%"
+				commandArgs = """
+					--target update-base-cache
+					--cache-from=registry.a8c.com/calypso/app:commit-${Settings.WpCalypso.paramRefs.buildVcsNumber},%base_image%
```

Comment:
cache-from seems to help the layer cache work better -- maybe worth trying for other things too!